File: src/product-entry-companions/internal.ts
import type {
  FamilyOrchestrationReferenceRef,
  JsonRecord,
  OperatorLoopActionSurface,
  ProductEntryProgressSurfaceInput,
  ProductEntryResumeContract,
  ProductEntryShellSurface,
  ProductEntryStartModeInput,
  ProductEntryStartResumeSurface,
  ProductEntryStepInput,
} from './types.ts';
import {
  validateFamilyDomainEntryContract as validateSharedFamilyDomainEntryContract,
  validateGatewayInteractionContract as validateSharedGatewayInteractionContract,
} from '../family-entry-contracts.ts';

export function isRecord(value: unknown): value is JsonRecord {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

export function optionalString(value: unknown) {
  return typeof value === 'string' && value.trim() ? value.trim() : null;
}

export function requireString(value: unknown, field: string) {
  const text = optionalString(value);
  if (!text) {
    throw new Error(`product entry companion 缺少字符串字段: ${field}`);
  }
  return text;
}

export function requireBoolean(value: unknown, field: string) {
  if (typeof value !== 'boolean') {
    throw new Error(`product entry companion 缺少布尔字段: ${field}`);
  }
  return value;
}

export function requireInteger(value: unknown, field: string) {
  if (!Number.isInteger(value)) {
    throw new Error(`product entry companion 缺少整数字段: ${field}`);
  }
  return Number(value);
}

export function requireRecord(value: unknown, field: string): JsonRecord {
  if (!isRecord(value)) {
    throw new Error(`product entry companion 缺少对象字段: ${field}`);
  }
  return value;
}

export function readOptionalStringProperty(value: JsonRecord, key: string, field: string) {
  if (!Object.prototype.hasOwnProperty.call(value, key)) {
    return undefined;
  }
  return requireString(value[key], field);
}

export function readStringList(value: unknown, field: string) {
  if (!Array.isArray(value)) {
    throw new Error(`product entry companion 缺少数组字段: ${field}`);
  }
  const normalized = value
    .map((entry, index) => requireString(entry, `${field}[${index}]`));
  return normalized;
}

export function optionalStringList(value: unknown, field: string) {
  if (!Array.isArray(value)) {
    return null;
  }
  return value.map((entry, index) => requireString(entry, `${field}[${index}]`));
}

export function normalizeResumeContract(value: unknown, field: string): ProductEntryResumeContract {
  const payload = requireRecord(value, field);
  return {
    surface_kind: requireString(payload.surface_kind, `${field}.surface_kind`),
    session_locator_field: requireString(payload.session_locator_field, `${field}.session_locator_field`),
    checkpoint_locator_field: optionalString(payload.checkpoint_locator_field),
  };
}

export function normalizeStep(value: unknown, field: string): ProductEntryStepInput {
  const payload = requireRecord(value, field);
  return {
    step_id: requireString(payload.step_id, `${field}.step_id`),
    title: requireString(payload.title, `${field}.title`),
    command: requireString(payload.command, `${field}.command`),
    surface_kind: requireString(payload.surface_kind, `${field}.surface_kind`),
    summary: requireString(payload.summary, `${field}.summary`),
    requires: readStringList(payload.requires, `${field}.requires`),
  };
}

export function normalizeProgressSurface(value: unknown, field: string): ProductEntryProgressSurfaceInput {
  const payload = requireRecord(value, field);
  return {
    surface_kind: requireString(payload.surface_kind, `${field}.surface_kind`),
    command: requireString(payload.command, `${field}.command`),
    step_id: optionalString(payload.step_id),
  };
}

export function normalizeStartMode(value: unknown, field: string): ProductEntryStartModeInput {
  const payload = requireRecord(value, field);
  return {
    mode_id: requireString(payload.mode_id, `${field}.mode_id`),
    title: requireString(payload.title, `${field}.title`),
    command: requireString(payload.command, `${field}.command`),
    surface_kind: requireString(payload.surface_kind, `${field}.surface_kind`),
    summary: requireString(payload.summary, `${field}.summary`),
    requires: readStringList(payload.requires, `${field}.requires`),
  };
}

export function normalizeStartResumeSurface(value: unknown, field: string): ProductEntryStartResumeSurface {
  const payload = requireRecord(value, field);
  const normalized: ProductEntryStartResumeSurface = {
    surface_kind: requireString(payload.surface_kind, `${field}.surface_kind`),
  };
  const command = readOptionalStringProperty(payload, 'command', `${field}.command`);
  if (command !== undefined) {
    normalized.command = command;
  }
  const sessionLocatorField = readOptionalStringProperty(
    payload,
    'session_locator_field',
    `${field}.session_locator_field`,
  );
  if (sessionLocatorField !== undefined) {
    normalized.session_locator_field = sessionLocatorField;
  }
  const checkpointLocatorField = readOptionalStringProperty(
    payload,
    'checkpoint_locator_field',
    `${field}.checkpoint_locator_field`,
  );
  if (checkpointLocatorField !== undefined) {
    normalized.checkpoint_locator_field = checkpointLocatorField;
  }
  return normalized;
}

export function normalizeProductEntryShellSurface(value: unknown, field: string): ProductEntryShellSurface {
  const payload = requireRecord(value, field);
  const normalized: ProductEntryShellSurface = {
    command: requireString(payload.command, `${field}.command`),
    surface_kind: requireString(payload.surface_kind, `${field}.surface_kind`),
  };
  const summary = readOptionalStringProperty(payload, 'summary', `${field}.summary`);
  if (summary !== undefined) {
    normalized.summary = summary;
  }
  const purpose = readOptionalStringProperty(payload, 'purpose', `${field}.purpose`);
  if (purpose !== undefined) {
    normalized.purpose = purpose;
  }
  const commandTemplate = readOptionalStringProperty(
    payload,
    'command_template',
    `${field}.command_template`,
  );
  if (commandTemplate !== undefined) {
    normalized.command_template = commandTemplate;
  }
  const requires = optionalStringList(payload.requires, `${field}.requires`);
  if (requires !== null) {
    normalized.requires = requires;
  }
  return {
    ...payload,
    ...normalized,
  };
}

export function normalizeOperatorLoopAction(value: unknown, field: string): OperatorLoopActionSurface {
  const payload = requireRecord(value, field);
  return {
    ...payload,
    command: requireString(payload.command, `${field}.command`),
    surface_kind: requireString(payload.surface_kind, `${field}.surface_kind`),
    summary: requireString(payload.summary, `${field}.summary`),
    requires: readStringList(payload.requires, `${field}.requires`),
  };
}

export function cloneRecord(value: unknown, field: string): JsonRecord {
  return { ...requireRecord(value, field) };
}

export const FRONTDESK_SHARED_HANDOFF_KEYS = [
  'direct_entry_builder',
  'opl_handoff_builder',
] as const;

export function normalizeFrontdeskSummary(value: unknown, field: string) {
  const payload = requireRecord(value, field);
  return {
    frontdesk_command: requireString(payload.frontdesk_command, `${field}.frontdesk_command`),
    recommended_command: requireString(payload.recommended_command, `${field}.recommended_command`),
    operator_loop_command: requireString(payload.operator_loop_command, `${field}.operator_loop_command`),
  };
}

export function mergeExtraPayload<T extends JsonRecord>(base: T, extraPayload: unknown, surfaceKind: string): T {
  if (extraPayload === undefined) {
    return base;
  }
  const normalizedExtraPayload = cloneRecord(extraPayload, 'extra_payload');
  for (const key of Object.keys(normalizedExtraPayload)) {
    if (Object.prototype.hasOwnProperty.call(base, key)) {
      throw new Error(`${surfaceKind} extra_payload 不允许覆盖核心字段: ${key}`);
    }
  }
  return {
    ...base,
    ...normalizedExtraPayload,
  } as T;
}

export function validateFamilyReferenceRef(value: unknown, field: string): FamilyOrchestrationReferenceRef {
  const payload = requireRecord(value, field);
  const normalized: FamilyOrchestrationReferenceRef = {
    ...payload,
    ref_kind: requireString(payload.ref_kind, `${field}.ref_kind`),
    ref: requireString(payload.ref, `${field}.ref`),
  };
  const label = readOptionalStringProperty(payload, 'label', `${field}.label`);
  if (label !== undefined) {
    normalized.label = label;
  }
  return normalized;
}

export function validateOptionalFamilyReferenceRef(value: unknown, field: string) {
  if (value === undefined) {
    return undefined;
  }
  return validateFamilyReferenceRef(value, field);
}

export function validateDomainEntryContractShape(value: unknown, field: string) {
  return validateSharedFamilyDomainEntryContract(value, field);
}

export function validateGatewayInteractionContractShape(value: unknown, field: string) {
  return validateSharedGatewayInteractionContract(value, field);
}

File: python/opl-harness-shared/src/opl_harness_shared/product_entry_companions/internal.py
from __future__ import annotations

from typing import Any, Mapping

from ..family_entry_contracts import (
    validate_family_domain_entry_contract as _validate_shared_family_domain_entry_contract,
    validate_gateway_interaction_contract as _validate_shared_gateway_interaction_contract,
    validate_shared_handoff as _validate_shared_handoff,
    validate_shared_handoff_builder as _validate_shared_handoff_builder,
)

_FRONTDESK_SHARED_HANDOFF_KEYS = ("direct_entry_builder", "opl_handoff_builder")


def _non_empty_text(value: object) -> str | None:
    text = str(value or "").strip()
    return text or None


def _require_string(value: object, field: str) -> str:
    text = _non_empty_text(value)
    if text is None:
        raise ValueError(f"product entry companion 缺少字符串字段: {field}")
    return text


def _require_bool(value: object, field: str) -> bool:
    if not isinstance(value, bool):
        raise ValueError(f"product entry companion 缺少布尔字段: {field}")
    return value


def _require_int(value: object, field: str) -> int:
    if not isinstance(value, int):
        raise ValueError(f"product entry companion 缺少整数字段: {field}")
    return value


def _require_mapping(value: object, field: str) -> Mapping[str, Any]:
    if not isinstance(value, Mapping):
        raise ValueError(f"product entry companion 缺少对象字段: {field}")
    return value


def _require_string_list(value: object, field: str) -> list[str]:
    if not isinstance(value, list):
        raise ValueError(f"product entry companion 缺少数组字段: {field}")
    return [_require_string(entry, f"{field}[{index}]") for index, entry in enumerate(value)]


def _optional_string_list(value: object, field: str) -> list[str] | None:
    if not isinstance(value, list):
        return None
    return [_require_string(entry, f"{field}[{index}]") for index, entry in enumerate(value)]


def _normalize_resume_contract(value: object, field: str) -> dict[str, str]:
    payload = _require_mapping(value, field)
    normalized = {
        "surface_kind": _require_string(payload.get("surface_kind"), f"{field}.surface_kind"),
        "session_locator_field": _require_string(
            payload.get("session_locator_field"),
            f"{field}.session_locator_field",
        ),
    }
    checkpoint_locator_field = _non_empty_text(payload.get("checkpoint_locator_field"))
    if checkpoint_locator_field is not None:
        normalized["checkpoint_locator_field"] = checkpoint_locator_field
    return normalized


def _normalize_step(value: object, field: str) -> dict[str, Any]:
    payload = _require_mapping(value, field)
    return {
        "step_id": _require_string(payload.get("step_id"), f"{field}.step_id"),
        "title": _require_string(payload.get("title"), f"{field}.title"),
        "command": _require_string(payload.get("command"), f"{field}.command"),
        "surface_kind": _require_string(payload.get("surface_kind"), f"{field}.surface_kind"),
        "summary": _require_string(payload.get("summary"), f"{field}.summary"),
        "requires": _require_string_list(payload.get("requires"), f"{field}.requires"),
    }


def _normalize_progress_surface(value: object, field: str) -> dict[str, Any]:
    payload = _require_mapping(value, field)
    normalized = {
        "surface_kind": _require_string(payload.get("surface_kind"), f"{field}.surface_kind"),
        "command": _require_string(payload.get("command"), f"{field}.command"),
    }
    step_id = _non_empty_text(payload.get("step_id"))
    if step_id is not None:
        normalized["step_id"] = step_id
    return normalized


def _read_optional_string_field(payload: Mapping[str, Any], key: str, field: str) -> str | None:
    if key not in payload:
        return None
    return _require_string(payload.get(key), field)


def _normalize_start_mode(value: object, field: str) -> dict[str, Any]:
    payload = _require_mapping(value, field)
    return {
        "mode_id": _require_string(payload.get("mode_id"), f"{field}.mode_id"),
        "title": _require_string(payload.get("title"), f"{field}.title"),
        "command": _require_string(payload.get("command"), f"{field}.command"),
        "surface_kind": _require_string(payload.get("surface_kind"), f"{field}.surface_kind"),
        "summary": _require_string(payload.get("summary"), f"{field}.summary"),
        "requires": _require_string_list(payload.get("requires"), f"{field}.requires"),
    }


def _normalize_start_resume_surface(value: object, field: str) -> dict[str, str]:
    payload = _require_mapping(value, field)
    normalized = {
        "surface_kind": _require_string(payload.get("surface_kind"), f"{field}.surface_kind"),
    }
    command = _read_optional_string_field(payload, "command", f"{field}.command")
    if command is not None:
        normalized["command"] = command
    session_locator_field = _read_optional_string_field(
        payload,
        "session_locator_field",
        f"{field}.session_locator_field",
    )
    if session_locator_field is not None:
        normalized["session_locator_field"] = session_locator_field
    checkpoint_locator_field = _read_optional_string_field(
        payload,
        "checkpoint_locator_field",
        f"{field}.checkpoint_locator_field",
    )
    if checkpoint_locator_field is not None:
        normalized["checkpoint_locator_field"] = checkpoint_locator_field
    return normalized


def _normalize_product_entry_shell_surface(value: object, field: str) -> dict[str, Any]:
    payload = _require_mapping(value, field)
    normalized: dict[str, Any] = {
        **dict(payload),
        "command": _require_string(payload.get("command"), f"{field}.command"),
        "surface_kind": _require_string(payload.get("surface_kind"), f"{field}.surface_kind"),
    }
    summary = _read_optional_string_field(payload, "summary", f"{field}.summary")
    if summary is not None:
        normalized["summary"] = summary
    purpose = _read_optional_string_field(payload, "purpose", f"{field}.purpose")
    if purpose is not None:
        normalized["purpose"] = purpose
    command_template = _read_optional_string_field(
        payload,
        "command_template",
        f"{field}.command_template",
    )
    if command_template is not None:
        normalized["command_template"] = command_template
    requires = _optional_string_list(payload.get("requires"), f"{field}.requires")
    if requires is not None:
        normalized["requires"] = requires
    return normalized


def _normalize_operator_loop_action(value: object, field: str) -> dict[str, Any]:
    payload = _require_mapping(value, field)
    return {
        **dict(payload),
        "command": _require_string(payload.get("command"), f"{field}.command"),
        "surface_kind": _require_string(payload.get("surface_kind"), f"{field}.surface_kind"),
        "summary": _require_string(payload.get("summary"), f"{field}.summary"),
        "requires": _require_string_list(payload.get("requires"), f"{field}.requires"),
    }


def _clone_mapping(value: object, field: str) -> dict[str, Any]:
    return dict(_require_mapping(value, field))


def _normalize_frontdoor_summary(value: object, field: str) -> dict[str, str]:
    payload = _require_mapping(value, field)
    return {
        "frontdoor_command": _require_string(payload.get("frontdoor_command"), f"{field}.frontdoor_command"),
        "recommended_command": _require_string(payload.get("recommended_command"), f"{field}.recommended_command"),
        "operator_loop_command": _require_string(
            payload.get("operator_loop_command"),
            f"{field}.operator_loop_command",
        ),
    }


def _normalize_frontdesk_summary(value: object, field: str) -> dict[str, str]:
    payload = _require_mapping(value, field)
    return {
        "frontdesk_command": _require_string(payload.get("frontdesk_command"), f"{field}.frontdesk_command"),
        "recommended_command": _require_string(payload.get("recommended_command"), f"{field}.recommended_command"),
        "operator_loop_command": _require_string(
            payload.get("operator_loop_command"),
            f"{field}.operator_loop_command",
        ),
    }


def _merge_extra_payload(base: dict[str, Any], extra_payload: object | None, *, surface_kind: str) -> dict[str, Any]:
    if extra_payload is None:
        return base
    normalized_extra_payload = _clone_mapping(extra_payload, "extra_payload")
    for key in normalized_extra_payload:
        if key in base:
            raise ValueError(f"{surface_kind} extra_payload 不允许覆盖核心字段: {key}")
    return {
        **base,
        **normalized_extra_payload,
    }


def _validate_family_reference_ref(value: object, field: str) -> dict[str, Any]:
    payload = _require_mapping(value, field)
    normalized = {
        **dict(payload),
        "ref_kind": _require_string(payload.get("ref_kind"), f"{field}.ref_kind"),
        "ref": _require_string(payload.get("ref"), f"{field}.ref"),
    }
    label = _read_optional_string_field(payload, "label", f"{field}.label")
    if label is not None:
        normalized["label"] = label
    return normalized


def _validate_optional_family_reference_ref(value: object, field: str) -> dict[str, Any] | None:
    if value is None:
        return None
    return _validate_family_reference_ref(value, field)


def _validate_domain_entry_contract_shape(value: object, field: str) -> dict[str, Any]:
    return _validate_shared_family_domain_entry_contract(value, field)


def _validate_gateway_interaction_contract_shape(value: object, field: str) -> dict[str, Any]:
    return _validate_shared_gateway_interaction_contract(value, field)

